src/renderer/render.rs

```diff
@@ -104,100 +104,74 @@ pub fn render_scene(
-        });
-
-        // Shadow items: z ordered
-        let mut shadow_items: Vec<(_, _)> = render_vectors
-            .iter()
-            .filter(|(_, v)| v.layer == Layer::Shadow)
-            .collect();
-        shadow_items.sort_by(|(_, a), (_, b)| {
-            let a = a.transform.translation().z;
-            let b = b.transform.translation().z;
-            a.partial_cmp(&b).unwrap_or(std::cmp::Ordering::Equal)
-        });
-        vector_render_queue.append(&mut shadow_items);
-
-        // Ground items: y ordered
-        let mut middle_items: Vec<(_, _)> = render_vectors
-            .iter()
-            .filter(|(_, v)| v.layer == Layer::Ground)
-            .collect();
-        middle_items.sort_by(|(_, a), (_, b)| {
-            let a = a.transform.translation().y;
-            let b = b.transform.translation().y;
-            b.partial_cmp(&a).unwrap_or(std::cmp::Ordering::Equal)
-        });
-        vector_render_queue.append(&mut middle_items);
-
-        // Foreground items: z ordered
-        let mut fg_items: Vec<(_, _)> = render_vectors
-            .iter()
-            .filter(|(_, v)| v.layer == Layer::Foreground)
-            .collect();
-        fg_items.sort_by(|(_, a), (_, b)| {
-            let a = a.transform.translation().z;
-            let b = b.transform.translation().z;
-            a.partial_cmp(&b).unwrap_or(std::cmp::Ordering::Equal)
-        });
-        vector_render_queue.append(&mut fg_items);
-
-        // UI items
-        let mut ui_items: Vec<(_, _)> = render_vectors
-            .iter()
-            .filter(|(_, v)| v.layer == Layer::UI)
-            .collect();
-        ui_items.sort_by(|(_, a), (_, b)| {
-            let a = a.transform.translation().z;
-            let b = b.transform.translation().z;
-            a.partial_cmp(&b).unwrap_or(std::cmp::Ordering::Equal)
-        });
-        vector_render_queue.append(&mut ui_items);
+        enum RenderItem<'a> {
+            Vector(&'a ExtractedRenderVector),
+            Text(&'a ExtractedRenderText),
+        }
+        let mut render_queue: Vec<(f32, RenderMode, (&PreparedAffine, RenderItem))> =
+            render_vectors
+                .iter()
+                .map(|(a, b)| {
+                    (
+                        b.transform.translation().z,
+                        b.render_mode,
+                        (a, RenderItem::Vector(b)),
+                    )
+                })
+                .collect();
+        render_queue.extend(query_render_texts.iter().map(|(a, b)| {
+            (
+                b.transform.translation().z,
+                b.render_mode,
+                (a, RenderItem::Text(b)),
+            )
+        }));
+
+        // Sort by render mode with screen space on top, then by z-index
+        render_queue.sort_by(
+            |(a_z_index, a_render_mode, _), (b_z_index, b_render_mode, _)| {
+                let z_index = a_z_index.partial_cmp(b_z_index).unwrap();
```

Comment:
This is concerning, would this cause a panic in the program if two f32s aren't comparable? (e.g. `f32::NaN` cannot be compared to `0.0`)

I think it would make more sense to default to Equal, e.g. `a.partial_cmp(&b).unwrap_or(std::cmp::Ordering::Equal)`

Reply:
Done.